### DISCUSSION TITLE
Elsa conditions

### DISCUSSION THREAD

**Discussion by AdcoConstruct** [elsa-workflows/elsa-core#2760] [Q&A]
I am trying to define an approval workflow. In the workflow I need to add an approver if a particular value is exceeded. The basis of flow is shown below.

User 1 approves
User 2 approves if a value exceeds a predetermined value.
User 3 approves

Below is some test code I've put together. If the condition is true, and User2 approves. However, the WriteHttpResponse with the User3 signal urls is not executed after the User2Join is executed.

Would anyone be able to help on this one?

.If(context => context.GetVariable<decimal>("ContractValue") > 250000
                , @if =>
                {
                    @if.When(OutcomeNames.True).WriteLine("Greater than 250000. STR required!")
                        .WriteHttpResponse(x =>
                        {
                            x.WithStatusCode(HttpStatusCode.OK);
                            x.WithContent(c =>
                                new Approval
                                {
                                    ApproverRole = ApproversRoles.User2,
                                    ApproveUrl = c.GenerateSignalUrl("User2Approve"),
                                    RejectUrl = c.GenerateSignalUrl("User2Reject")
                                }
                            );
                        })
                        .Then<Fork>(activity => activity.WithBranches("User2Approve", "User2Reject"), fork =>
                        {
                            fork
                                .When("User2Approve")
                                .SignalReceived("User2Approve")
                                .WriteLine("User2 Approved")
                                .ThenNamed("User2Join");
                            fork
                                .When("User2Reject")
                                .SignalReceived("User2Reject")
                                .WriteLine("User2 Rejected")
                                .ThenNamed("User2Join");
                        })
                        .Add<Join>(join => join.WithMode(Join.JoinMode.WaitAny)).WithName("User2Join");
                }
            )
            .WriteHttpResponse(x =>
            {
                x.WithStatusCode(HttpStatusCode.OK);
                x.WithContent(c =>
                    new Approval
                    {
                        ApproverRole = ApproversRoles.User3, ApproveUrl = c.GenerateSignalUrl("User3Approve"),
                        RejectUrl = c.GenerateSignalUrl("User3Reject")
                    }
                );
            })
            .Then<Fork>(activity => activity.WithBranches("User3Approve", "User3Reject"), fork =>
            {
                fork
                    .When("User3Approve")
                    .SignalReceived("User3Approve")
                    .WriteLine("User3 Approved")
                    .ThenNamed("User3Join");
                fork
                    .When("User3Reject")
                    .SignalReceived("User3Reject")
                    .WriteLine("User3 Rejected")
                    .ThenNamed("User3Join");
            })
            .Add<Join>(join => join.WithMode(Join.JoinMode.WaitAny)).WithName("User3Join")

**Comment by AdcoConstruct** [comment]
I have persisted with trying to find a solution for this issue with no luck. I really like the library and it's capabilities but I can't get it to fit my requirements. Before I give up on this, would anyone be able to provide an example of how I can build an approval workflow that requires approvals from a specific person if a condition is met. For example User1 approves, User2 approve if condition is true, User 3 approves. It seems that the done path is not executing after the condition logic is complete.

**Reply by sfmskywalker** [reply]
Can you share your example so I can take a look at it? Your sample code snippet looks OK so I am curious to see why it doesn't follow the path you think it should.

**Reply by AdcoConstruct** [reply]
[Workflow example.zip](https://github.com/elsa-workflows/elsa-core/files/8128508/Workflow.example.zip)

Hi, I have attached an example. The workflow path is **/v1/testworkflow**

I am sending  in the following payload

{
 "lettingvalue" : 400000
}

When the CFO role approval is provided, the HTTP response for the CM approval isn't received. The workflow is then finalised.

Any help on this is very much appreciated. Thanks

**Reply by AdcoConstruct** [reply]
Hi @sfmskywalker, have you had an opportunity to look at my code sample? I need to make a decision on whether to roll my own workflow. However, I really want to use this library as believe this save me a great deal of time, but I'm stuck. Cheers

**Reply by sfmskywalker** [reply]
Not yet, but I’ll take a look into it tomorrow and will let you know my findings.

**Reply by AdcoConstruct** [reply]
Many thanks.

**Reply by sfmskywalker** [reply]
I can confirm that the **Done** branch of the **If** activity is not executed after its **True** branch completes:

![image](https://user-images.githubusercontent.com/938393/156149324-4fdfac35-3afe-4248-80b8-537caa4f7b3e.png)

This is a bug that I need to look into.

**Reply by sfmskywalker** [reply]
I created #2794

**Reply by AdcoConstruct** [reply]
Thanks for the update @sfmskywalker

**Reply by sfmskywalker** [reply]
I just merged [this PR](https://github.com/elsa-workflows/elsa-core/pull/2797) that fixes the bug and will be available through the 2.6.0-preview packages [available from MyGet](https://elsa-workflows.github.io/elsa-core/docs/next/installation/installing-feeds#myget) once [this build](https://ci.appveyor.com/project/sfmskywalker/elsa/builds/42744785) completes.

**Comment by AdcoConstruct** [comment]
This is great. Many thanks for your efforts in getting this resolved. Cheers


Andy Fordyce
Application Developer
[https://www.adcoconstruct.com.au/images/Transparent.png]

02 8437 5000
0438 398 162
Level 2, 7-9 West Street
North Sydney NSW 2060
adcoconstruct.com.au


[https://www.adcoconstruct.com.au/images/ADCO_Email_sig.png]


[https://www.adcoconstruct.com.au/images/Transparent.png]
[https://www.adcoconstruct.com.au/images/Line_transparent.png]
[https://www.adcoconstruct.com.au/images/Transparent.png]
[https://www.adcoconstruct.com.au/wp-content/uploads/2021/11/ADCO-2021-Deloitte-Best-Managed-Companies-Email-Signature.jpg]<https://www2.deloitte.com/au/en/pages/deloitte-private/articles/best-managed-companies.html>
[https://www.adcoconstruct.com.au/images/Transparent.png]

This email is strictly confidential. If you are not the addressee indicated in this message (or responsible for delivery of the message to such person), you may not copy or deliver this message to anyone. In such case, you should destroy this message and notify us immediately. If you or your employer does not consent to email messages of this kind, please advise us immediately. ADCO does not guarantee the integrity of emails or attached files. The views or opinions expressed are the author’s own and may not reflect the views or opinions of ADCO.

**Reply by mubinahmed** [reply]
@AdcoConstruct 
Hi Andy, do you have the JSON file of the above workflow? I am designing exactly same workflow and it will be helpful for my understanding.
